### ISSUE TITLE
Remove `rich` as a core dependencies

### ISSUE THREAD

**Issue by noklam** [cookiecutter/cookiecutter#2078]
### Description:
https://github.com/cookiecutter/cookiecutter/pull/1901, as far as I can see, `rich` is being added as a core dependencies but I cannot find any discussion before that.

While `rich` is useful for local development, it causes a lot of issues outside of standard terminal and is not useful in non-interactive environment (i.e. CI/production etc).

This make `cookiecutter` even bigger in size. For example, `fastapi-cli` has a slim version `fastapi-cli-slim` to remove `rich` from its dependencies list.

The original change move from `click.prompt` to `rich.Prompt` , essentially this function.
```python
def read_user_variable
```

However, it's not possible to get around it without monkey-patching. 

### Alternative/Implementation
- Fallback to non-rich Prompts when `rich` is not installed
- Remove `rich` from core dependencies and move it to optional instead.

**Comment by JamesParrott** [comment]
Is cookiecutter a performance critical tool for many users?  Does rich add much more than about 1.6MB?  

```
(rich) C:\...\venvs>pip install rich
Collecting rich
  Using cached rich-13.7.1-py3-none-any.whl (240 kB)
Collecting markdown-it-py>=2.2.0 (from rich)
  Using cached markdown_it_py-3.0.0-py3-none-any.whl (87 kB)
Collecting pygments<3.0.0,>=2.13.0 (from rich)
  Using cached pygments-2.18.0-py3-none-any.whl (1.2 MB)
Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich)
  Using cached mdurl-0.1.2-py3-none-any.whl (10.0 kB)
Installing collected packages: pygments, mdurl, markdown-it-py, rich
Successfully installed markdown-it-py-3.0.0 mdurl-0.1.2 pygments-2.18.0 rich-13.7.1

[notice] A new release of pip is available: 23.1.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip

```

**Comment by noklam** [comment]
@JamesParrott Thanks for your reply. I want to make it clear that my suggestion is to make an "option" to install `cookiecutter` without `rich`. While filesize is a factor to consider, it is not the most important one here.

1. Dependencies footprint
I do think `rich` has done a pretty good job at keeping its dependencies footprint small. Adding new dependencies always add burden to the maintenance.

2. Transient problem from `rich`
This is IMO the biggest problem. It is a common pattern of the Python community to disable rich by `pip uninstall rich`, where the libraries will try to detect if `rich` is importable and automatically fallback to the standard CLI. By forcing `rich` as a core dependencies, `cookiecutter` cannot be used together with the "slim" version that the community has developed.

This is our main issue, **we want to provide an option that is "rich-free"**, but `cookiecutter` is part of our core dependencies. As a result, we have to consider to remove `cookiecutter` or we are stuck in the older version of `cookiecutter` which are both sub-optimal. We are relying on the same fallback mechanism to use `rich` when it's available. We want to use `cookiecutter` as a library but not so much with the CLI/prompt.

The alternative is forcing other libraries to have rich installed but not using it, which IMO is not a good option.

**Comment by JamesParrott** [comment]
You're welcome.  Has the community developed a slim version of cookiecutter yet?  I can't find a fork.  It's straightforward enough on the surface to add a `try` `except` around this line:

```
from rich.prompt import Confirm, InvalidResponse, Prompt, PromptBase
```
and if it errors, patch in classes that point to the old CLI.  But rich is also then used in 5 tests too (I've not updated my clone, there could be more).  And cookiecutter workflow would need to test both versions.

Have you considered using jinja2 on the jinja2 templates directly in Kedro, and removing cookiecutter?

**Comment by noklam** [comment]
> Has the community developed a slim version of cookiecutter yet? I can't find a fork. 

No, this is exactly why I open this issue and hoping that the maintainers agree with it. Either a silm version or just move `rich` to optional (IMO optional is better in this case since rich is such a small component here, it's not worth the effort to maintain a separate package)

>  It's straightforward enough on the surface to add a try except around this line:

The fix wouldn't be too complicated like you said, but it has to be done in `cookiecutter` and we don't want to do any monkeypatch from our side.


> Have you considered using jinja2 on the jinja2 templates directly in Kedro, and removing cookiecutter?

For sure, we have considered alternative like `copier` etc. I don't think `Jinja2` can replace `cookiecutter` for us since we are also using some more advance feature in `cookiecutter` like "post init hook"

**Comment by k1shk1n** [comment]
I'd like to point out that there are problems on the side of how much it weighs, too. If you factor in rich and its dependencies, it goes something like this:
```
$ du -sh .venv
16M     .venv

$ source .venv/bin/activate
(.venv) $ pip install rich --no-cache-dir
Collecting rich
  Downloading rich-14.0.0-py3-none-any.whl.metadata (18 kB)
Collecting markdown-it-py>=2.2.0 (from rich)
  Downloading markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)
Collecting pygments<3.0.0,>=2.13.0 (from rich)
  Downloading pygments-2.19.2-py3-none-any.whl.metadata (2.5 kB)
Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich)
  Downloading mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)
Downloading rich-14.0.0-py3-none-any.whl (243 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.2/243.2 kB 940.7 kB/s eta 0:00:00
Downloading markdown_it_py-3.0.0-py3-none-any.whl (87 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.5/87.5 kB 6.1 MB/s eta 0:00:00
Downloading pygments-2.19.2-py3-none-any.whl (1.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 4.8 MB/s eta 0:00:00
Downloading mdurl-0.1.2-py3-none-any.whl (10.0 kB)
Installing collected packages: pygments, mdurl, markdown-it-py, rich
Successfully installed markdown-it-py-3.0.0 mdurl-0.1.2 pygments-2.19.2 rich-14.0.0

(.venv) $ du -sh .venv/
28M     .venv/
```
12 mb is so many for one prompting..

even if we use `uv` (its virtual environment comes out smaller), we get the following:
```
$ uv venv
Using CPython 3.11.11
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate

$ du -sh .venv
88K     .venv

$ source .venv/bin/activate
(.venv) $ uv pip install rich --no-cache-dir
Resolved 4 packages in 1.46s
Prepared 4 packages in 648ms
Installed 4 packages in 74ms
 + markdown-it-py==3.0.0
 + mdurl==0.1.2
 + pygments==2.19.2
 + rich==14.0.0

(.venv) $ du -sh .venv
6.7M    .venv
```
6.6 mb still a lot

maybe we can do something with it or use some of the things @noklam suggested
